File: daily-coding-problems/2-strings/2-1-find-anagram-indices/java/solution.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {

  interface anonFunc {
    int[] makeFreq(String w);
  }

  public static int[] solution(String w, String s) {
    int size = s.length();
    int a = 'a';

    anonFunc func = new anonFunc() {
      public int[] makeFreq(String w) {
        int[] res = new int[26];
        for(int c : w.toLowerCase().toCharArray()) {
          res[c - a]++;
        }
        return res;
      }
    };

    int[] ref = func.makeFreq(w);
    int[] curr = func.makeFreq("");
    ArrayList<Integer> result = new ArrayList<>();

    for(int end = 0; end < size; end++) {
      int start = end - w.length() + 1;

      int endC = s.charAt(end);
      curr[endC - a]++;

      if (Arrays.equals(ref, curr)) {
        result.add(start);
      }

      if (start >= 0) {
        int startC = s.charAt(start);
        curr[startC - a]--;
      }
    }

    return result.stream().mapToInt(i -> i).toArray();
  }

  public static void main(String[] args) {
    String w = "ab";
    String s = "abxaba";
    System.out.println("w: " + w);
    System.out.println("s: " + s);

    System.out.println("solution: " + Arrays.toString(solution(w, s)));
  }
}

File: daily-coding-problems/2-strings/2-1-find-anagram-indices/scala/solution.scala
import scala.collection.mutable

object Solution {
  def solution(w: String, s: String): List[Int] = {
    val a = 'a'.toInt
    val size = s.length

    def makeMapping(w: String): Array[Int] = {
      var res = new Array[Int](26)
      w.foreach(c => {
        res(c.toInt - a) += 1
      })
      return res
    }

    val ref = makeMapping(w)
    var curr = makeMapping("")
    var result = new mutable.ListBuffer[Int]()

    for (end <- 0 to (size - 1)) {
      val endC = s.charAt(end).toInt
      val start = end - w.length + 1

      curr(endC - a) += 1

      if (ref.sameElements(curr)) {
        result += start
      }

      if (start >= 0) {
        val startC = s.charAt(start).toInt
        curr(startC - a) -= 1
      }
    }

    return result.toList
  }

  def main(args: Array[String]): Unit = {
    val w = "ab"
    val s = "abxaba"
    println("w: " + w)
    println("s: " + s)

    println("solution: " + solution(w, s))
  }
}
